lib/common/common.go

```diff
@@ -0,0 +1,350 @@
+
+		if layerData.Architecture != "" {
+			arch, _ := appctypes.NewACName("arch")
+			parentLabels = append(parentLabels, appctypes.Label{Name: *arch, Value: layerData.Architecture})
+		}
+	}
+
+	genManifest.Labels = labels
+
+	if dockerConfig != nil {
+		exec := getExecCommand(dockerConfig.Entrypoint, dockerConfig.Cmd)
+		if exec != nil {
+			user, group := parseDockerUser(dockerConfig.User)
+			var env appctypes.Environment
+			for _, v := range dockerConfig.Env {
+				parts := strings.SplitN(v, "=", 2)
+				env.Set(parts[0], parts[1])
+			}
+			app := &appctypes.App{
+				Exec:             exec,
+				User:             user,
+				Group:            group,
+				Environment:      env,
+				WorkingDirectory: dockerConfig.WorkingDir,
+			}
+			genManifest.App = app
+		}
+	}
+
+	if layerData.Parent != "" {
+		var dependencies appctypes.Dependencies
+		parentAppNameString := dockerURL.IndexURL + "/" + dockerURL.ImageName + "-" + layerData.Parent
+		parentAppNameString, err := appctypes.SanitizeACName(parentAppNameString)
+		if err != nil {
+			return nil, err
+		}
+		parentAppName, err := appctypes.NewACName(parentAppNameString)
+		if err != nil {
+			return nil, err
+		}
+
+		dependencies = append(dependencies, appctypes.Dependency{App: *parentAppName, Labels: parentLabels})
+
+		genManifest.Dependencies = dependencies
+	}
+
+	return genManifest, nil
+}
+
+func writeACI(layer io.ReadSeeker, manifest schema.ImageManifest, curPwl []string, output string) (*schema.ImageManifest, error) {
+	aciFile, err := os.Create(output)
+	if err != nil {
+		return nil, fmt.Errorf("error creating ACI file: %v", err)
+	}
+	defer aciFile.Close()
+
+	gw := gzip.NewWriter(aciFile)
+	defer gw.Close()
+	trw := tar.NewWriter(gw)
+	defer trw.Close()
+
+	var whiteouts []string
+	convWalker := func(t *tarball.TarFile) error {
+		name := t.Name()
+		if name == "./" {
+			return nil
+		}
+		t.Header.Name = path.Join("rootfs", name)
+		absolutePath := strings.TrimPrefix(t.Header.Name, "rootfs")
+		if strings.Contains(t.Header.Name, "/.wh.") {
+			whiteouts = append(whiteouts, strings.Replace(absolutePath, ".wh.", "", -1))
+			return nil
+		}
+		if t.Header.Typeflag == tar.TypeLink {
+			t.Header.Linkname = path.Join("rootfs", t.Linkname())
+		}
+
+		if err := trw.WriteHeader(t.Header); err != nil {
+			return err
+		}
+		if _, err := io.Copy(trw, t.TarStream); err != nil {
+			return err
+		}
+
+		if !util.In(curPwl, absolutePath) {
+			curPwl = append(curPwl, absolutePath)
+		}
+
+		return nil
+	}
+	reader, err := aci.NewCompressedTarReader(layer)
+	if err == nil {
+		// Write files in rootfs/
+		if err := tarball.Walk(*reader, convWalker); err != nil {
+			return nil, err
+		}
+	} else {
+		// ignore errors
+	}
+	newPwl := subtractWhiteouts(curPwl, whiteouts)
+
+	manifest.PathWhitelist = newPwl
+	if err := addMinimalACIStructure(trw, manifest); err != nil {
+		return nil, fmt.Errorf("error writing rootfs entry: %v", err)
```

Comment:
This is slightly misleading since this also writes the manifest. Although I am wondering if it would be better to write the rootfs dir before doing the convWalker anyway..